src/_nebari/stages/kubernetes_keycloak/__init__.py

```diff
@@ -315,6 +318,414 @@ def _attempt_keycloak_connection(
+                    realm_name="master",
+                    client_id="admin-cli",
+                    verify=False,
+                )
+
+                # Check if nebari-bot already exists
+                users = admin.get_users({"username": "nebari-bot"})
+
+                if users:
+                    print("nebari-bot user already exists")
+                    user_id = users[0]["id"]
+
+                    # Reset password to ensure it matches the expected value
+                    # (Keycloak doesn't allow reading passwords for comparison)
+                    admin.set_user_password(
+                        user_id=user_id, password=nebari_bot_password, temporary=False
+                    )
+                    print("Updated nebari-bot password to match expected value")
+                else:
+                    # Create nebari-bot user
+                    user_id = admin.create_user(
+                        {
+                            "username": "nebari-bot",
+                            "enabled": True,
+                            "credentials": [
+                                {
+                                    "type": "password",
+                                    "value": nebari_bot_password,
+                                    "temporary": False,
+                                }
+                            ],
+                        }
+                    )
+                    print("Successfully created nebari-bot user")
+
+                # Assign admin role to nebari-bot user
+                # Get the admin role from master realm
+                admin_role = admin.get_realm_role("admin")
+
+                # Check if user already has the admin role
+                user_roles = admin.get_realm_roles_of_user(user_id)
+                has_admin_role = any(role.get("name") == "admin" for role in user_roles)
+
+                if not has_admin_role:
+                    admin.assign_realm_roles(user_id, [admin_role])
+                    print("Assigned admin role to nebari-bot user")
+                else:
+                    print("nebari-bot user already has admin role")
+
+                # Success - break out of retry loop
+                break
+
+            except KeycloakError as e:
+                if attempt < max_attempts:
+                    print(f"Attempt {attempt}/{max_attempts} failed: {e}")
+                    print(f"Retrying in {retry_delay} seconds...")
+                    time.sleep(retry_delay)
+                else:
+                    print(
+                        f"Failed to configure nebari-bot user after {max_attempts} attempts: {e}"
+                    )
+                    sys.exit(1)
+
+    def _restore_keycloak_database(self, backup_file: Path):
+        """Restore PostgreSQL database from backup file using Kubernetes exec."""
+        import base64
+        import tarfile
+        from io import BytesIO
+        from pathlib import Path
```

Comment:
This is already imported at the upper level.

Reply:
moved